topic_tools/test/test_topic_tool.hpp

```diff
@@ -176,3 +184,64 @@ class TestTopicToolMultiSub : public TestTopicTool
   std::string target_input_topic_prefix_;
   std::string target_output_topic_;
 };
+
+class TestTopicToolMultiPub : public TestTopicTool
+{
+public:
+  void SetUp()
+  {
+    using std::placeholders::_1;
+    const std::string test_name = ::testing::UnitTest::GetInstance()->current_test_info()->name();
+    test_node_ = rclcpp::Node::make_shared(test_name);
+    target_input_topic_ = "/" + test_name + "/input";
+    target_output_topic_prefix_ = "/" + test_name + "/output";
+    publisher_ = test_node_->create_publisher<std_msgs::msg::String>(target_input_topic_, 10);
+    std::vector<std::string> topic_names = get_target_output_topics();
+    set_number_of_subscriptions(topic_names.size());
+    for (size_t i = 0; i < topic_names.size(); i++) {
+      subscriptions_[i] = test_node_->create_subscription<std_msgs::msg::String>(
+        topic_names[i], 10,
+        [this, i](std_msgs::msg::String::ConstSharedPtr msg) {this->topic_callback(msg, i);});
+    }
+  }
+
+  void TearDown()
+  {
+    test_node_.reset();
+    publisher_.reset();
+    for (int i = 0; i < num_target_output_topics_; i++) {
+      subscriptions_[i].reset();
+    }
+  }
+
+  void publish_and_check(std::string msg_content, std::shared_ptr<rclcpp::Node> target_node)
+  {
+    auto message = std_msgs::msg::String();
+    message.data = msg_content;
+    publisher_->publish(message);
+    rclcpp::spin_some(target_node);
+    rclcpp::spin_some(test_node_);
```

Comment:
This was following the pattern from [other](https://github.com/rcywongaa/topic_tools/blob/add_demux/topic_tools/test/test_topic_tool.hpp#L97-L98)  [tests](https://github.com/rcywongaa/topic_tools/blob/add_demux/topic_tools/test/test_topic_tool.hpp#L158-L159).
I believe spinning nodes continuously in a separate thread would be similarly non-deterministic.

Would you mind to elaborate a bit more on the `spin_until_future_complete` way?  I don't think there's a way to get a `future` out of the `publish()` atm?

I think this boils down to the asynchronous nature of `publish`.  The discussion about a synchronous version of `publish()` [here](https://github.com/ros2/ros2/issues/255) and [eProsima FastDDS's synchronous publishing](https://discourse.ros.org/t/rmw-fast-dds-publication-mode-sync-vs-async-and-how-to-change-it/17153) might be relevant

In particular, the following section from [FastDDS](https://www.eprosima.com/index.php/resources-all/performance/dds-asynchronous-vs-synchronous-publishing)
> Additionally, if there are local readers (Intraprocess), the data is directly sent to each one (no transport is involved), their listeners are notified, and the corresponding callbacks are executed before returning control to the user.

seems like what we need for having deterministic callbacks.

Reply:
Ahh never mind. There is too much to rework, and since other tests follow the same design pattern in this package, let's go with the current solution. 
Will rework in a follow up PR if CI will be unstable.

As regards 
> Would you mind to elaborate a bit more on the spin_until_future_complete way? I don't think there's a way to get a future out of the publish() atm?

I meant that the future could be an async task where we could check the number of messages that arrived during the timeout, for example.

A good example of how to orchestrate such a helper class for subscriptions can be found in the rosbag2 https://github.com/ros2/rosbag2/blob/rolling/rosbag2_test_common/include/rosbag2_test_common/subscription_manager.hpp